userspace/engine/rule_loader.h

```diff
@@ -33,110 +33,207 @@ class falco_engine;
+		falco_engine* engine;
+	};
+
+	/*!
+		\brief Represents infos about an engine version requirement
+	*/
+	struct engine_version_info
+	{
+		uint32_t version;
+	};
+
+	/*!
+		\brief Represents infos about a plugin version requirement
+	*/
+	struct plugin_version_info
+	{
+		std::string name;
+		std::string version;
+	};
 
 	/*!
-		\brief Returns the rules loaded after the last invocation of load()
+		\brief Represents infos about a list 
 	*/
-	virtual const indexed_vector<falco_rule>& rules();
+	struct list_info
+	{
+		mark context;
+		bool used;
+		size_t index;
+		size_t visibility;
+		std::string name;
+		std::vector<std::string> items;
+	};
 
 	/*!
-		\brief Configures the loader. The changes will influence the next
-		invocation of load().
-		\param min_priority The minimum priority below which rules are skipped
-		by the loader
-		\param extra Text to be appended/substituted in the output of all rules
-		\param replace_container_info If true, the extra string is used to
-		replace the "%container.info" token in rules outputs. If false, the
-		"%container.info" token is substituted with a default text and the
-		extra string is appended at the end of the rule output. If a rule
-		output does not contain "%container.info", then this flag has no effect
-		and the extra string is appended at the end of the rule output anyways.
+		\brief Represents infos about a macro 
 	*/
-	virtual void configure(
-		falco_common::priority_type min_priority,
-		bool replace_container_info,
-		const std::string& extra);
+	struct macro_info
+	{
+		mark context;
+		bool used;
+		size_t index;
+		size_t visibility;
+		std::string name;
+		std::string cond;
+		std::string source;
+		std::shared_ptr<libsinsp::filter::ast::expr> cond_ast;
+	};
+
+	/*!
+		\brief Represents infos about a single rule exception
+	*/
+	struct rule_exception_info
+	{
+		/*!
+			\brief This is necessary due to the dynamic-typed nature of
+			exceptions. Each of fields, comps, and values, can either be a
+			single value or a list of values. This is a simple hack to make
+			this easier to implement in C++, that is not non-dynamic-typed.
+		*/
+		struct entry {
+			bool is_list;
+			std::string item;
+			std::vector<entry> items;
+
+			inline bool is_valid() const
+			{
+				return (is_list && !items.empty())
+					|| (!is_list && !item.empty());
+			}
+		};
+
+		std::string name;
+		entry fields;
+		entry comps;
+		std::vector<entry> values;
+	};
+
+	/*!
+		\brief Represents infos about a rule 
+	*/
+	struct rule_info
+	{
+		mark context;
+		size_t index;
+		size_t visibility;
+		std::string name;
+		std::string cond;
+		std::string source;
+		std::string desc;
+		std::string output;
+		std::set<std::string> tags;
+		std::vector<rule_exception_info> exceptions;
+		falco_common::priority_type priority;
+		bool enabled;
+		bool warn_evttypes;
+		bool skip_if_unknown_filter;
+	};
+
+	/*!
+		\brief Erases all the internal state and definitions
+	*/
+	virtual void clear();
```

Comment:
Why are a bunch of methods virtual? I checked and I didn't see any derived class from rule_loader. Did I miss it?

Reply:
I've put them as virtual so that it can be easier to write unit tests by mocking the loader. IIRC we use FakeIt here, which needs virtual methods to create a Mock.

Reply:
Ah ok, sounds good.